### PR TITLE
[macOS] Handle interleaved movement and gesture events

### PR DESCRIPTION
**Pull request by knopp**
Fixes https://github.com/flutter/flutter/issues/146893

The crash happens, because with magic mouse, it is quite common to receive `mouseExited:` event while the gesture event is still in progress. This would reset the gesture state while the gesture event is still active, leading to assertion when the gesture event ends. 

The state machine assumes that when mouse exists that also ends any gesture events. However that is not the case with magic mouse, which handles the movement and gesture events separately.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by knopp** [comment]
@moffatman, can you think of any unwanted side-effects of not calling `GestureReset()` from `Reset()`?

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm [hunk 1]
```diff
@@ -1004,6 +1011,39 @@ - (bool)testTrackpadGesturesAreSentToFramework:(id)engineMock {
   return true;
 }
 
+- (bool)mouseAndGestureEventsAreHandledSeparately:(id)engineMock {
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:1014]: Optional: Might be worth adding a comment mentioning Apple's Magic Mouse.

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

File: shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm (hunk 1 above)
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:1014]: It is mentioned inside the test but that's probably too easy to overlook.

Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm:1014]: Ah yes -- somehow glossed over that on line 1034!

**Review by moffatman — APPROVED** [review]
Thx

**Review by cbracken — COMMENTED** [review]
Oh wow the lint bot upset about something that's not formatting :) Looks like the linter caught a couple leaks -- CFRelease the values allocated by CGEventCreateScrollWheelEvent and CGEventCreateCopy.

**Pull request merged**